### PR TITLE
RPC: `compositeSwap`

### PR DESCRIPTION
**Pull request by ivan-zynesis**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
FC new RPC

#### Which issue(s) does this PR fixes?:

Fixes #

#### Additional comments?:
#488

### CODE REVIEW

**Review by jingyi2811 — DISMISSED** [review]

File: packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+import { UTXO } from 'packages/jellyfish-api-core/src/category/wallet'
+import BigNumber from 'bignumber.js'
+
+describe('compositeSwap', () => {
+  const container = new LoanMasterNodeRegTestContainer()
+  const testing = Testing.create(container)
+  const client = new ContainerAdapterClient(container)
+
+  beforeAll(async () => {
+    await container.start()
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts:16]:
```suggestion
    await testing.container.start()
```

File: packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+    await container.waitForWalletCoinbaseMaturity()
+    await testing.token.dfi({ amount: 30000 })
+    await container.generate(1)
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts:19]:
```suggestion
    await testing.generate(1)
```

File: packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+    })
+    await testing.poolpair.add({
+      a: { symbol: 'XYZ', amount: 50 },
+      b: { symbol: 'PATHB', amount: 1 }
+    })
+    await container.generate(1)
+  })
+
+  afterAll(async () => {
+    await container.stop()
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts:85]:
```suggestion
    await testing.container.stop()
```

File: packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+      maxPrice: 1.2
+    }
+
+    const hex = await client.poolpair.compositeSwap(metadata)
+    expect(typeof hex).toStrictEqual('string')
+    expect(hex.length).toStrictEqual(64)
+    await container.generate(1)
+
+    { // after swap
+      const fromBalances = await client.account.getAccount(fromAddress)
```
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/poolpair/compositeSwap.test.ts:155]:
```suggestion
      const fromBalances = await testing.rpc.account.getAccount(fromAddress)
```

**Review by canonbrother — DISMISSED** [review]

**Review by surangap — COMMENTED** [review]
looks good. Can it support composite swaps more than 2 levels?
Can we add a test for 3 levels?
i.e. A-B, B-C, C-D, and swap from A -> D
I still need to spend some time understanding the composite swap CPP code. It has some logic.

**Comment by ivan-zynesis** [comment]
> Can it support composite swaps more than 2 levels?

Yes, it should support from my understanding, the actually execution is 2 steps.
1. find the right path
2. execute multiple swap based on pool ids
thus it should work regardless how long the path is

>  If yes, Can we add a test for 3 levels? i.e. A-B, B-C, C-D, and swap from A -> D
Sure

**Review by jingyi2811 — APPROVED** [review]

**Review by surangap — APPROVED** [review]

**Review by fuxingloh — APPROVED** [review]

**Pull request merged**